From 528fc27885e315b6f69d429e0f57f534059c118c Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Mon, 11 Apr 2022 11:35:01 +0200
Subject: [PATCH 0126/1695] plat-stm32mp2: sysconfig: Handle IO compensation

Add interface to enable and disable IO compensation
for the five IO domains.
Low power has to be handled by the user.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git d7a96e6457b22d574025e80af10cbf89179cf625 plat-stm32mp2: sysconfig: Handle IO compensation]
Change-Id: I09b32920a4b0424f7c0dfe86e719e2cad2d87d79
Signed-off-by: Pascal Paillet <p.paillet@st.com>
---
 .../plat-stm32mp2/drivers/stm32mp25_syscfg.c  | 83 +++++++++++++++++++
 core/arch/arm/plat-stm32mp2/stm32_sysconf.h   | 15 ++++
 2 files changed, 98 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_syscfg.c b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_syscfg.c
index 71e24ab58..e19197e80 100644
--- a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_syscfg.c
+++ b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_syscfg.c
@@ -5,6 +5,7 @@
 
 #include <initcall.h>
 #include <io.h>
+#include <kernel/delay.h>
 #include <mm/core_memprot.h>
 #include <stdbool.h>
 #include <stm32_sysconf.h>
@@ -15,9 +16,34 @@
 #define SYSCON_BANK(id)     (((id) & GENMASK_32(31, 16)) >> 16)
 
 struct io_pa_va syscfg_base[SYSCON_NB_BANKS] = {
+	{ .pa = SYSCFG_BASE },
 	{ .pa = A35SSC_BASE }
 };
 
+/* IO compensation CCR registers bit definition */
+#define SYSCFG_CCCR_CS				BIT(9)
+#define SYSCFG_CCCR_EN				BIT(8)
+#define SYSCFG_CCCR_RAPSRC_MASK			GENMASK_32(7, 4)
+#define SYSCFG_CCCR_RANSRC_MASK			GENMASK_32(3, 0)
+
+/* IO compensation CCSR registers bit definition */
+#define SYSCFG_CCSR_READY			BIT(8)
+#define SYSCFG_CCSR_APSRC_MASK			GENMASK_32(7, 4)
+#define SYSCFG_CCSR_ANSRC_MASK			GENMASK_32(3, 0)
+
+#define SYSCFG_CCSR_READY_TIMEOUT_US		U(1000)
+
+/*
+ * SYSCFG IO compensation register offsets (base relative)
+ */
+static uint32_t syscfg_cccr_offset[SYSFG_NB_IO_ID] = {
+	[SYSFG_VDDIO1_ID] = SYSCON_ID(SYSCON_SYSCFG, 0x4020U),
+	[SYSFG_VDDIO2_ID] = SYSCON_ID(SYSCON_SYSCFG, 0x4018U),
+	[SYSFG_VDDIO3_ID] = SYSCON_ID(SYSCON_SYSCFG, 0x4000U),
+	[SYSFG_VDDIO4_ID] = SYSCON_ID(SYSCON_SYSCFG, 0x4008U),
+	[SYSFG_VDD_IO_ID] = SYSCON_ID(SYSCON_SYSCFG, 0x4010U),
+};
+
 static vaddr_t stm32mp_syscfg_base(uint32_t id)
 {
 	uint32_t bank = SYSCON_BANK(id);
@@ -38,3 +64,60 @@ uint32_t stm32mp_syscfg_read(uint32_t id)
 {
 	return io_read32(stm32mp_syscfg_base(id) + SYSCON_OFFSET(id));
 }
+
+TEE_Result stm32mp25_syscfg_enable_io_compensation(enum syscfg_io_ids id)
+{
+	vaddr_t cccr_addr = stm32mp_syscfg_base(SYSCON_SYSCFG) +
+			    SYSCON_OFFSET(syscfg_cccr_offset[id]);
+	vaddr_t ccsr_addr = cccr_addr + 4U;
+	uint64_t timeout_ref = 0;
+
+	FMSG("Enable IO comp for id %u", id);
+
+	if (io_read32(ccsr_addr) & SYSCFG_CCSR_READY)
+		return TEE_SUCCESS;
+
+	io_setbits32(cccr_addr, SYSCFG_CCCR_EN);
+
+	timeout_ref = timeout_init_us(SYSCFG_CCSR_READY_TIMEOUT_US);
+
+	while (!(io_read32(ccsr_addr) & SYSCFG_CCSR_READY))
+		if (timeout_elapsed(timeout_ref))
+			break;
+
+	if (!(io_read32(ccsr_addr) & SYSCFG_CCSR_READY)) {
+		EMSG("IO compensation cell not ready");
+		return TEE_ERROR_GENERIC;
+	}
+
+	io_clrbits32(cccr_addr, SYSCFG_CCCR_CS);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result stm32mp25_syscfg_disable_io_compensation(enum syscfg_io_ids id)
+{
+	vaddr_t cccr_addr = stm32mp_syscfg_base(SYSCON_SYSCFG) +
+			    SYSCON_OFFSET(syscfg_cccr_offset[id]);
+	vaddr_t ccsr_addr = cccr_addr + 4U;
+	uint32_t value = 0;
+
+	FMSG("Disable IO comp for id %u", id);
+
+	if (!(io_read32(ccsr_addr) & SYSCFG_CCSR_READY))
+		return TEE_SUCCESS;
+
+	value = io_read32(ccsr_addr) & (SYSCFG_CCSR_APSRC_MASK |
+		                        SYSCFG_CCSR_ANSRC_MASK);
+
+	io_clrsetbits32(cccr_addr,
+			SYSCFG_CCCR_RAPSRC_MASK | SYSCFG_CCCR_RANSRC_MASK,
+			value);
+
+	io_setbits32(cccr_addr, SYSCFG_CCCR_CS);
+
+	io_clrbits32(cccr_addr, SYSCFG_CCCR_EN);
+
+	return TEE_SUCCESS;
+}
+
diff --git a/core/arch/arm/plat-stm32mp2/stm32_sysconf.h b/core/arch/arm/plat-stm32mp2/stm32_sysconf.h
index 109db3467..7dc09ebe6 100644
--- a/core/arch/arm/plat-stm32mp2/stm32_sysconf.h
+++ b/core/arch/arm/plat-stm32mp2/stm32_sysconf.h
@@ -9,10 +9,12 @@
 #include <assert.h>
 #include <stdbool.h>
 #include <stdint.h>
+#include <tee_api_types.h>
 #include <types_ext.h>
 
 /* syscon banks */
 enum syscon_banks {
+	SYSCON_SYSCFG,
 	SYSCON_A35SSC,
 	SYSCON_NB_BANKS
 };
@@ -42,4 +44,17 @@ enum syscon_banks {
 void stm32mp_syscfg_write(uint32_t id, uint32_t value, uint32_t bitmsk);
 uint32_t stm32mp_syscfg_read(uint32_t id);
 
+/* IO comp identifiers */
+enum syscfg_io_ids {
+	SYSFG_VDDIO1_ID,
+	SYSFG_VDDIO2_ID,
+	SYSFG_VDDIO3_ID,
+	SYSFG_VDDIO4_ID,
+	SYSFG_VDD_IO_ID,
+	SYSFG_NB_IO_ID
+};
+
+TEE_Result stm32mp25_syscfg_enable_io_compensation(enum syscfg_io_ids id);
+TEE_Result stm32mp25_syscfg_disable_io_compensation(enum syscfg_io_ids id);
+
 #endif /*__STM32_SYSCONF_H__*/
-- 
2.47.2

